### ISSUE TITLE
tfd.MultivariateNormalTriL does not behave as expected in a transition function for MarkovChain

### ISSUE THREAD

**Issue by xiaolongluo1979** [tensorflow/probability#1389]
Hi, I am trying to use MultivariateNormal distribution  to construct a Markov chain with multidimensional state (in each component). While the simple tfd.MultivariateNormalDiag works fine, the more realistic one, tfd.MultivariateNormalTriL does not work as expected for multiple samples (sample() is fine but sample (2) runs into "weird" errors). Can someone help to see how it can be fixed? Thanks.

### These codes work well:
batch_size=5
dim_YC=3

previous_state={}
previous_state['time']=tf.random.uniform(shape=[batch_size])
previous_state['YC0']=tf.random.uniform(shape=[batch_size,dim_YC])

### define transition model
jumptime_parameters=tf.keras.layers.Dense(1,activation='relu', name='jumptime_b')
YC0_loc=tf.keras.layers.Dense(dim_YC,activation='relu', name='YC0_loc')

def transition_fn0(_, previous_state):
    Lstates= [previous_state[v] for v in previous_state.keys() if v!='time']
    tfstate=tf.concat([tf.cast(tf.expand_dims(previous_state['time'],axis=-1)
                               ,dtype=tf.float32)]
                      +Lstates,axis=-1)
    return tfd.JointDistributionNamedAutoBatched(
      batch_ndims=tf.rank(previous_state['time']), 
      model=dict(
        time = tfd.TransformedDistribution(
                  distribution=tfd.Gamma(
                        concentration=tf.constant(1.0,dtype=tf.float32), 
                        rate=tf.math.softplus(jumptime_parameters(tfstate)[...,-1])),
            bijector=tfp.bijectors.Shift(previous_state['time'])),
        YC0 = lambda time:tfd.MultivariateNormalDiag(
                loc=YC0_loc(tf.concat([tfstate, tf.expand_dims(time,axis=-1)],
                                      axis=-1)), 
                scale_diag=dim_YC*[0.001]),
    ))

dist0=transition_fn0(_, previous_state)
dist0.log_prob(dist0.sample(2))

### These codes replaces tfd.MultivariateNormalDiag by tfd.MultivariateNormalTriL with tfp.util.TransformedVariable but won't work for multiple samples

batch_size=5
dim_YC=3
previous_state={}
previous_state['time']=tf.random.uniform(shape=[batch_size])
previous_state['YC']=tf.random.uniform(shape=[batch_size,dim_YC])

### define transition model
jumptime_parameters=tf.keras.layers.Dense(1,activation='relu', name='jumptime_b')
YC_loc=tf.keras.layers.Dense(dim_YC,activation='relu', name='YC_loc')

def transition_fn(_, previous_state):
    Lstates= [previous_state[v] for v in previous_state.keys() if v!='time']
    tfstate=tf.concat([tf.cast(tf.expand_dims(previous_state['time'],axis=-1)
                               ,dtype=tf.float32)]
                      +Lstates,axis=-1)
    return tfd.JointDistributionNamedAutoBatched(
      batch_ndims=tf.rank(previous_state['time']), 
      model=dict(
        time = tfd.TransformedDistribution(
                  distribution=tfd.Gamma(
                        concentration=tf.constant(1.0,dtype=tf.float32), 
                        rate=tf.math.softplus(jumptime_parameters(tfstate)[...,-1])),
            bijector=tfp.bijectors.Shift(previous_state['time'])),
        YC = lambda time:tfd.MultivariateNormalTriL(
        loc=YC_loc(tf.concat([tfstate, tf.expand_dims(time,axis=-1)],
                             axis=-1)),
        scale_tril=tfp.util.TransformedVariable(
            tf.eye(dim_YC, dtype=tf.float32),
            tfp.bijectors.FillScaleTriL(),
            name="raw_scale_tril"))
    ))

dist=transition_fn(_, previous_state)

dist.log_prob(dist.sample()) ### work ok

dist.log_prob(dist.sample(2))  ### not work

**Comment by xiaolongluo1979** [comment]
@davmre Hi Dave, you solved a similar(?) issue related to tfd.TransformedDistribution. Is there a similar issue with tfp.util.TransformedVariable ? Thanks.

**Comment by davmre** [comment]
I think the issue here is that you're trying to create a variable inside of a model function. You'd get an error doing that with MVNDiag as well (and the issue isn't specific to `TransformedVariable` either, a plain `tf.Variable` would cause the same problem). You only want the variable to be created once, and reused every time the model runs, so you need to create it outside of the model:

```python
scale_tril_var = tfp.util.TransformedVariable(tf.eye(dim_YC, dtype=tf.float32),
                                              tfp.bijectors.FillScaleTriL(),
                                              name="raw_scale_tril")

def transition_fn(_, previous_state):
  # ... 
  return tfd.JointDistributionNamedAutoBatched(
    # ...
      scale_tril=scale_tril_var)
  )
```

**Comment by xiaolongluo1979** [comment]
@davmre Thanks for the comment. Yes, it does work with the variable created it outside of the model! Alternatively, i also tried to fill in the triangular with keras model parameters (defined outside the model) bypassing tfp.util.TransformedVariable, which seemed to work as well. Any comment on the comparison between the two approaches? Which is more robust for tfp? Thanks.

scale_tril=0.001+tfp.math.fill_triangular(
tf.keras.layers.Dense(int(n_YC*(n_YC+1)/2),activation='relu', name='YC_scale')(
tf.concat([tfstate, tf.expand_dims(time,axis=-1)],
                                         axis=-1)))

**Comment by xiaolongluo1979** [comment]
resolved
